### ISSUE TITLE
LWS_CALLBACK_CLIENT_RECEIVE_PONG not received after updating to new versions

### ISSUE THREAD

**Issue by sp-stefano-lanza** [warmcat/libwebsockets#1271]
Hi,

I am in charge of updating our client to the latest lws version. The client has its own strategy to track the state of a connection, which works by pinging the server (LWS_WRITE_PING) and by receiving pongs via the LWS_CALLBACK_CLIENT_RECEIVE_PONG callback. That works fine with version 2.0.3 of lws.
After updating to version 2.4.2 or 3.0.0, we noticed that the client never receives  the LWS_CALLBACK_CLIENT_RECEIVE_PONG. Do you have an explanation for this breaking change ? The only ping-related change in the release logs appear to be the new ws_ping_pong_interval  in lws_context_creation_info, which I set to 0. Setting that to another value does not fix this issue anyway. Has LWS_CALLBACK_CLIENT_RECEIVE_PONG been deprecated after lws introduced its own ping tracking mechanism ?
Any help with this issue would be very much appreciated. Thanks!

**Comment by lws-team** [comment]
I added a minimal-ws-client-ping example to confirm it, on both master and v3.0-stable.

https://github.com/warmcat/libwebsockets/tree/master/minimal-examples/ws-client/minimal-ws-client-ping

There doesn't seem to be any problem...

**Comment by sp-stefano-lanza** [comment]
First, I bow to you for the amazing support.
By comparing your example code with mine, I think I found the cause of the issue. Our client sends ping messages with a payload of 0 bytes:
lws_write(wsi, data + LWS_PRE, 0, LWS_WRITE_PING);
In version 2.0.3 of lws, this works and the client gets a corresponding pong. In recent versions of lws there seems to be some filtering done internally which suppresses the LWS_CALLBACK_CLIENT_RECEIVE_PONG callback. Charles still shows the ping-pong exchange btw. 
If instead I send some bytes of data along with the ping message, then lws correctly notifies the client with a pong callback.
Is this a bug in lws or is it the expected behavior ?

**Comment by lws-team** [comment]
I see... well I can reproduce that... I added a '-z' flag to the minimal example to send a zero-length ping.

It sends it OK and receives it but as you say it filters it... in one place it's specifically filtering zero-length PONG and in another place it's accidentally filtering it if extensions aren't enabled.

I pushed a patch just on master to see what happens to Autobahn tests that run in travis with the filtering removed, it has many tests around zero length this and that, mixed in between various tricky things like multibyte utf-8 etc.  If it's still happy I'll keep the change and put it on v3.0-stable, there wasn't any plan on my part to hide zero length pings.  It takes an hour or so to do all the tests.
